Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm

```diff
@@ -102,4 +131,41 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
     EXPECT_FALSE(didReceiveQueryPermission);
 }
 
+TEST(PermissionsAPI, PermissionMayHaveChanged)
```

Comment:
In general, we try to avoid API tests unless they are strictly necessary. Why aren't we using layout tests here?
Seems like the existing testRunner.grantWebNotificationPermission() / denyWebNotificationPermission() could cause permissionChanged() to be called, then we should be able to use layout tests.

The reason we prefer API tests is that they are faster to run (can be parallelized) and can be more easily skipped.

Reply:
I think API test is used to ensure WebKit API works correctly; layout test is used to ensure Web API work correctly.
testRunner.grantWebNotificationPermission() is used to simulate client's behavior in a layout test. We are not testing API in layout test; we are assuming it works. And we won't be able to check how many times queryPermission is called in layout test.

We should probably adopt permissionChanged in grantWebNotificationPermission/denyWebNotificationPermission though, to make TestRunner act as a functional client.

Reply:
> we won't be able to check how many times queryPermission is called in layout test
Does this API test really check that?

> I think API test is used to ensure WebKit API works correctly; layout test is used to ensure Web API work correctly.
Here the API is really here to support a Web feature though: The firing of a change event.

If you want an API test to validate that the API behaves correctly, and that queryPermission is only called once, fine. However, this needs layout testing too then.
Future tests involving workers, dedicated workers and service workers should definitely be layout tests and not API tests.

Reply:
The change event is a Web feature, not a native API feature. Ideally we'd be using WPT tests with a WebDriver hook to change the permissions.

Reply:
Yeah the patch has both WebKit API and Web API changes, so we should probably add both tests